submissions/kotlyar-andrey/frogger-game/app.js

```diff
@@ -0,0 +1,249 @@
+  this._player = player;
+  this.start();
+}
+Enemy.prototype = Object.create(ImageSprite.prototype);
+
+Enemy.prototype.getRow = function () {
+  return Math.ceil((this._y - 60) / CELL_HEIGHT) + 1;
+};
+
+Enemy.prototype.update = function (dt) {
+  this._x += this._speed * dt;
+  if (this._x > SCREEN_WIDTH) {
+    this.start();
+  }
+  this.checkLose();
+};
+
+Enemy.prototype.start = function () {
+  this._x = -CELL_WIDTH;
+  this._speed =
+    MIN_ENEMY_SPEED +
+    Math.floor(Math.random() * (MAX_ENEMY_SPEED - MIN_ENEMY_SPEED));
+};
+
+Enemy.prototype.checkLose = function () {
+  const [eRow, eCol, pRow, pCol] = [
+    this.getRow(),
+    this.getColumn(),
+    this._player.getRow(),
+    this._player.getColumn(),
+  ];
+  if (eRow === pRow && (eCol === pCol || eCol - 1 === pCol)) {
+    this._player.lose();
+  }
+};
+
+function WinText(x, y, player) {
+  TextSprite.call(this, x, y, "green");
+  this._player = player;
+}
+WinText.prototype = Object.create(TextSprite.prototype);
+
+WinText.prototype.update = function () {
+  this._text = `Wins: ${this._player._wins}`;
+};
+
+function LoseText(x, y, player) {
+  TextSprite.call(this, x, y, "red");
+  this._player = player;
+}
+LoseText.prototype = Object.create(TextSprite.prototype);
+
+LoseText.prototype.update = function () {
+  this._text = `Loses: ${this._player._loses}`;
+};
+
+function StreakText(x, y, player) {
+  TextSprite.call(this, x, y, "gold");
+  this._player = player;
+}
+StreakText.prototype = Object.create(TextSprite.prototype);
+
+StreakText.prototype.update = function () {
+  this._text = `Streak: ${this._player._streak}`;
+};
+
+function JokeText(x, y) {
+  TextSprite.call(this, x, y, "white");
+  this._messages = {
+    first: "DON'T PRESS 'F' ( ͡ಠ ͜ʖ ͡ಠ)",
+    last: "AAHHAH, GOOD LUCK ✌",
+    nope: "NOPE ≧◉◡◉≦",
+  };
+  this._joke = this._messages.first;
+  this._counter = 0;
+  this._state = "normal";
+}
+JokeText.prototype = Object.create(TextSprite.prototype);
+
+JokeText.prototype.update = function () {
+  this._counter = (this._counter + 1) % 250;
+  this._text =
+    this._counter < 200 || this._state === "rotated" ? this._joke : "";
+};
+JokeText.prototype.jokeHandler = function () {
+  if (this._state === "normal") {
+    this._state = "rotated";
+    this._joke = this._messages.last;
+    this.rotateScreen();
+  } else {
+    this._joke = this._messages.nope;
+    setTimeout(() => {
+      this._joke = this._messages.last;
+    }, 1200);
+  }
+};
+JokeText.prototype.rotateScreen = function () {
+  ctx.translate(SCREEN_WIDTH, SCREEN_HEIGHT);
+  ctx.rotate(Math.PI);
+};
+
+/* Основная программа */
+document.addEventListener("keyup", function (e) {
+  var allowedKeys = {
+    37: "left",
+    38: "up",
+    39: "right",
+    40: "down",
+    70: "f",
+  };
+
+  player.handleInput(allowedKeys[e.keyCode]);
+});
+
+const SCREEN_WIDTH = 505;
+const SCREEN_HEIGHT = 606;
+const START_X = 202;
+const START_Y = 404;
+const CELL_WIDTH = 101;
```

Comment:
I feel these numbers are connected somehow...
As a peer developer tasked to add somethif to this project (e.g. field parameters) I'd like to change as few things as possible to implement changes.